src/init.cpp

```diff
@@ -1467,7 +1467,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     if (mempool_opts.max_size_bytes < 0 || mempool_opts.max_size_bytes < descendant_limit_bytes) {
         return InitError(strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0)));
     }
-    LogPrintf("* Using %.1f MiB for in-memory UTXO set (plus up to %.1f MiB of unused mempool space)\n", cache_sizes.coins * (1.0 / 1024 / 1024), mempool_opts.max_size_bytes * (1.0 / 1024 / 1024));
+    // If we are running in blocksonly mode don't allocate the default 300MB of space for mempool
+    // unless the user specifies -maxmempool, as otherwise it will leak into dbcache allowance.
+    // Keeping the mempool enabled will still permit transaction relay for whitelisted peers
+    if (args.GetBoolArg("-blocksonly", 0) && !args.IsArgSet("-maxmempool")) {
+        // Re-use descendant_limit_bytes as the minimum mempool size
+        mempool_opts.max_size_bytes = descendant_limit_bytes;
```

Comment:
nit: `descendant_limit_bytes` is around 3.9MB and not 5MB as claimed. Should the documentation be changed to reflect this or it doesn't really matter?